Fix doc.resolve for pages with <base> path

jotto opened the pull request:
when `this.url` is `/some/dir`
and HTML is: `<base href="/"><a href="hello">hello</a>`
  * before: `/some/dir/hello`
  * after: `/hello`

## N.B.

I just edited this inline on Github, I didn't clone/branch and test it locally.

jotto committed:
Fix doc.resolve for pages with <base> path

when `this.url` is `/some/dir`
and HTML is: `<base href="/"><a href="hello">hello</a>`
  * before: `/some/dir/hello`
  * after: `/hello`

flesler commented:
I have 2 thoughts about this:
1. As far as I remember (not too far), `<base>` only affects elements within it. Your version applies the `<base>` to all links on the page. For it to work as expected, one should instead provide the `<a>` element and the code should traverse it upwards to see if there's a `<base>`. Seems to hurt the usability of this method.
2. Seems to me (could be wrong) that `<base>` elements are very rare. Given that you are not forced to use this method, then if you are among the lucky ones that has to interact with these outliers, you can implement a better `resolve()` or "overload" the one provided by default.

What do you think?

jotto committed:
Don't join base+href if href starts with http

without this, url.resolve will happily create http://example.com/http://example.com/hello

jotto commented:
> As far as I remember (not too far), <base> only affects elements within it.

According to MDN: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/base

> There can be only one <base> element in a document. 

> If multiple <base> elements are specified, only the first href and first target value are used; all others are ignored.

but that link is the extent of my research

jotto commented:
> you can implement a better resolve() or "overload" the one provided by default.

I agree, but overriding is impossible since `Document` is not exposed... right? (unless we do a `require.cache[require.resolve("node-spider")]` hack).

So then 3 options:

1. expose the `Document` on the exported `Spider` function
2. add new method `resolveWithBase`
3. do nothing

jotto committed:
Fix var name

jotto commented:
## For posterity

in case this we don't add this functionality, here's a helper:

```javascript
var resolveUrl = function(doc, uri) {
  var base = doc.$('base').attr('href');
  if (base && !uri.match(/^(http|\/)/i)) uri = base + uri;

  return stdlibUrl.resolve(doc.url, uri);
}
```

### Usage

```javascript
var handleRequest = function(doc) {
  // new page crawled
  console.log(doc.res); // response object
  console.log(doc.url); // page url
  // uses cheerio, check its docs for more info
  doc.$('a').each(function(i, elem) {
      // do stuff with element
      var href = doc.$(elem).attr('href').split('#')[0];
      var url = resolveUrl(doc, href);
      // crawl more
      spider.queue(url, handleRequest);
  });
}
```

jotto committed:
Fix for pages with <base> and a leading slash

This mirrors how Chrome handles links with leading slash (they ignore the base href)

jotto commented:
It appears Chrome doesn't care where you put the `<base>` (putting it below a link still affects the path of that link)

![image](https://cloud.githubusercontent.com/assets/16573/23937281/e2432398-0913-11e7-817c-edd97a971ff6.png)
